include/swift/AST/DiagnosticEngine.h

```diff
@@ -629,6 +634,42 @@ namespace swift {
     DiagnosticState(DiagnosticState &&) = default;
     DiagnosticState &operator=(DiagnosticState &&) = default;
   };
+
+  /// Abstract interface class for providing diagnostics from YAML files.
+  class YAMLDiagnostic {
+    virtual void getDiagnostics(std::string locale, std::string path) = 0;
+  public:
+    virtual ~YAMLDiagnostic() { }
+  };
+
+  class YAMLDiagnosticProvider : public YAMLDiagnostic {
+    /// Diagnostic messages language code.
+    std::string diagnosticLocaleCode = "";
+    
+    /// Diagnostic messages directory path.
+    std::string diagnosticMessagesPath = "";
+  
+  public:
+    std::vector<DiagnosticNode> diagnostics;
+    explicit YAMLDiagnosticProvider(std::string locale, std::string path);
+    void getDiagnostics(std::string locale, std::string path) override;
+
+    std::string getDiagnosticLocaleCode() {
+      return diagnosticLocaleCode;
+    }
+    std::string getDiagnosticMessagesPath() {
+      return diagnosticMessagesPath;
+    }
+  };
+
+  class DiagnosticProducer {
+  public:
+    /// If the diagnostic message isn't available/localized in the current `yaml` file,
+    /// return the diagnostic message from the `.def` file.
+    std::string
+    getMessageOr(DiagID id,
+                 Optional<YAMLDiagnosticProvider> localization) const;
```

Comment:
Let's rename this to `LocalizationProducer` maybe that would be clearer what the intention is here, and second parameter of this method should be a fallback "default" diagnostic from `.def` file.

Reply:
by the "default" fallback you mean the message from the `.def` file right?

Reply:
Yes, that's right.